### ISSUE TITLE
Question: fast huffman decoder

### ISSUE THREAD

**Issue by Nomis101** [libjpeg-turbo/libjpeg-turbo#420]
There is a patch for libjpeg-turbo floating around in the web in the mozilla and google repo, which says "_This change ... is unlikely to be accepted upstream into libjpeg-turbo._" I was wondering, why this is unlikely to be accepted upstream?
https://hg.mozilla.org/mozilla-central/file/tip/media/libjpeg/1050342.diff#l1

**Comment by dcommander** [comment]
I’ll dig up the original discussion and post a link to it, but if memory serves, the patch caused a significant performance regression, and the sole purpose of the patch was to produce deterministic and canonical behavior (“canonical”=like libjpeg) for a certain set of corrupt JPEG images. My position was that libjpeg-turbo should behave like libjpeg for valid images, but in the spirit of GIGO (garbage in/garbage out), corrupt images should not crash libjpeg-turbo, but they can’t be guaranteed to always produce deterministic output or the same output as libjpeg, particularly if doing so would cause a performance regression.

**Comment by Nomis101** [comment]
Thanks for the clarification. 👍🏻

**Comment by dcommander** [comment]
Here's the thread: https://bugzilla.mozilla.org/show_bug.cgi?id=1050342.  Apparently the performance regression was only for 32-bit code, so I'll revisit this when I have a chance, since we're using newer compilers these days that might not regress as much, and since 32-bit code isn't as important anymore.

**Comment by dcommander** [comment]
I can confirm that at least part of the issue still exists with the latest libjpeg-turbo code base, although I can now only reproduce it with **jpeg-image-003.jpg** in **broken-images.tar.gz**.  I tried the 1.5.x code base as well, and I could still only repro the issue with that one image.  I can also confirm that the patch from Mozilla works around the issue.  Unfortunately, I just don't have the project resources to look into the performance right now.  It would take a while to re-run all of the benchmarks and establish confidence that there are no regressions.  Leaving this issue open until I can make the time to do that or perhaps find a better solution.

**Comment by dcommander** [comment]
Actually, I ended up having to do a quick & dirty benchmark run anyhow, in order to validate the performance neutrality of some undefined behavior fixes in the Huffman decoder.  Thus, I implemented a more minimal patch to work around the issue: https://github.com/libjpeg-turbo/libjpeg-turbo/commit/8fa70367ed5bbdc0685494e8212d975b5c89a77d. With this patch, I can no longer reproduce the issue with any of the test images in **broken-images.tar.gz**, and I can confirm that the patch is performance-neutral relative to 2.1 beta1.  I would appreciate it if the Mozilla and/or Chrome developers could check my work.
